crates/ruma-macros/src/events/event_enum.rs

```diff
@@ -293,51 +307,69 @@ fn expand_into_full_event(
 /// Create a content enum from `EventEnumInput`.
 fn expand_content_enum(
     kind: EventKind,
-    events: &[(LitStr, Path)],
+    events: &[EventEnumEntry],
+    docs: &[TokenStream],
     attrs: &[Attribute],
     variants: &[EventEnumVariant],
     ruma_common: &TokenStream,
-) -> TokenStream {
+) -> syn::Result<TokenStream> {
     let serde = quote! { #ruma_common::exports::serde };
     let serde_json = quote! { #ruma_common::exports::serde_json };
 
     let ident = kind.to_content_enum();
 
     let event_type_enum = kind.to_event_type_enum();
 
-    let content: Vec<_> =
-        events.iter().map(|(name, path)| to_event_content_path(kind, name, path, None)).collect();
-    let event_type_match_arms = events.iter().map(|(s, _)| {
-        if let Some(prefix) = s.value().strip_suffix(".*") {
-            quote! { _s if _s.starts_with(#prefix) }
-        } else {
-            quote! { #s }
+    let mut content = Vec::with_capacity(events.len());
+    let mut event_type_match_arms = Vec::with_capacity(events.len());
+    for (i, event) in events.iter().enumerate() {
+        let stable_name = event.stable_name()?;
+        let variant = &variants[i];
```

Comment:
This will increase MSRV to 1.59, but I think a very nice way to write this is a `use std::iter::zip;` plus

```suggestion
    for (event, variant) in zip(events, variants) {
        let stable_name = event.stable_name()?;
```